mavros_extras/src/plugins/joy_control.cpp

```diff
@@ -0,0 +1,237 @@
+/**
+ * @brief JoyControl plugin
+ * @file joy_control.cpp
+ * @author Francois Chataigner <chataign@gmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2016 Francois Chataigner.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <pluginlib/class_list_macros.h>
+
+#include <sensor_msgs/Joy.h>
+
+#include <mavros_msgs/SetMode.h>
+#include <mavros_msgs/CommandBool.h>
+#include <mavros_msgs/State.h>
+#include <mavros_msgs/SetMode.h>
+
+using namespace std;
+
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+namespace mavplugin {
+struct Control
+{
+	int axis;
+	double scale;
+	double offset;
+	double deadzone;
+
+	double map_axis( const sensor_msgs::Joy::ConstPtr& joy ) const
+	{
+		if ( axis < 0 || axis >= joy->axes.size() ) throw std::out_of_range("failed to map axis");
+		double val = joy->axes[axis] + offset;
+		return ( val > deadzone || val < -deadzone ? val * scale : 0.0 );
+	}
+};
+
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * @brief Joystick Control plugin
+ */
+class JoyControlPlugin : public MavRosPlugin
+{
+	ros::NodeHandle mavros_nh_, plugin_nh_;
+	UAS *uas_;
+	mavros_msgs::State state_;
+	mavlink_message_t ctrl_msg_;
+
+	Control x_, y_, z_, r_;
+
+	int arm_button_;
+	vector<int> mode_mapping_;
+	vector<string> mode_names_;
+
+	ros::Timer send_msg_timer_;
+	ros::ServiceClient arming_srv_, set_mode_srv_;
+	ros::Subscriber joystick_sub_, state_sub_;
+
+public:
+	JoyControlPlugin()
+		: mavros_nh_("~")
+		, plugin_nh_("~joy_control")
+		, uas_(nullptr) {}
+
+	void initialize(UAS &uas )
+	{
+		uas_ = &uas;
+
+		int msg_rate;
+
+		plugin_nh_.param<int>("msg_rate", msg_rate, 10 );
+		plugin_nh_.param<int>("arm_button", arm_button_, 2 );
```

Comment:
You do not need to specify type in each `NodeHandle::param<>()`, for most calls C++ type out works fine.